### PR TITLE
Report error on unicode BOM before #!

### LINKED ISSUE #205: 7$: Report error on BOM before #!
BOM: Byte Order Mark
#!: Shebang

Linux (and I assume macOS and other POSIX-like platforms) only interpret shebangs if the first byte in the file is `#` and the second byte in the file is `!` (I think; please confirm). This means that the presence of a BOM (UTF-8-encoded) will cause the shebang characters to not be interpreted by the operating system. This is almost definitely a mistake.

Report an error if this mistake occurs.

### PR DESCRIPTION
**Pull request by singalhimanshu**
Fixes #205

### CODE REVIEW

File: test/test-lex.cpp [hunk 1]
```diff
@@ -1876,6 +1876,18 @@ TEST_F(test_lex, lex_not_shebang) {
...
 
+  // BOM must not appear before '#!'.
+  {
+    error_collector v;
+    padded_string input(u8"\ufeff#!notashebang\n"_sv);
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::end_of_file) << "# should be skipped";
+
+    EXPECT_THAT(v.errors, ElementsAre(ERROR_TYPE_FIELD(
+                              error_unexpected_bom_before_shebang, where,
```
Comment (strager) [inline test/test-lex.cpp:1887]: Rename `where` to `bom`.

File: test/test-lex.cpp [hunk 2]
```diff
@@ -1876,6 +1876,18 @@ TEST_F(test_lex, lex_not_shebang) {
                               offsets_matcher(&input, 2, 3))));
   }
 
+  // BOM must not appear before '#!'.
```
Comment (strager) [inline test/test-lex.cpp:1879]: Make this a separate `TEST` and put this documentation in the test's name.

File: test/test-lex.cpp [hunk 3]
```diff
@@ -1876,6 +1876,18 @@ TEST_F(test_lex, lex_not_shebang) {
...
+  {
```
Comment (strager) [inline test/test-lex.cpp:1880]: Also test BOM without `#!`. (It should be ignored, producing no error.)

File: test/test-lex.cpp [hunk 4]
```diff
@@ -1876,6 +1876,18 @@ TEST_F(test_lex, lex_not_shebang) {
...
+    error_collector v;
+    padded_string input(u8"\ufeff#!notashebang\n"_sv);
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::end_of_file) << "# should be skipped";
+
+    EXPECT_THAT(v.errors, ElementsAre(ERROR_TYPE_FIELD(
+                              error_unexpected_bom_before_shebang, where,
+                              offsets_matcher(&input, 0, 2))));
```
Comment (strager) [inline test/test-lex.cpp:1888]: Blocking: Shouldn't the error cover all three bytes of the BOM, not just the first two bytes?

Write the test like this, which should make the issue more obvious:
```suggestion
                              offsets_matcher(&input, 0, u8"\ufeff"))));
```

**Review by strager — CHANGES_REQUESTED** [review]
Looks good aside from the weird error span.

File: test/test-lex.cpp (hunk 1 above)
Comment (singalhimanshu) [reply test/test-lex.cpp:1887]: Done

File: test/test-lex.cpp (hunk 2 above)
Comment (singalhimanshu) [reply test/test-lex.cpp:1879]: Shouldn't this belong to - https://github.com/quick-lint/quick-lint-js/blob/6c54dc042d691bbbe6d0c83115b5cd6e907c92bd/test/test-lex.cpp#L1842 as the name suggest `lex_not_shebang`. Any particular reason to move it as a separate `TEST`?

File: test/test-lex.cpp (hunk 3 above)
Comment (singalhimanshu) [reply test/test-lex.cpp:1880]: I guess there it already a test for that here - https://github.com/quick-lint/quick-lint-js/blob/6c54dc042d691bbbe6d0c83115b5cd6e907c92bd/test/test-lex.cpp#L1836

File: test/test-lex.cpp (hunk 4 above)
Comment (singalhimanshu) [reply test/test-lex.cpp:1888]: yeah makes sense

File: test/test-lex.cpp (hunk 2 above)
Comment (strager) [reply test/test-lex.cpp:1879]: > Any particular reason to move it as a separate `TEST`?

Yes. You thought the test was different enough to write a comment, so I think it's different enough to put into its own `TEST` block.

Also, all of the other tests are expecting an `error_unexpected_hash_character`, implying that the `#!` was not parsed as a comment. Your test is parsing the `#!` as a comment.

File: test/test-lex.cpp (hunk 3 above)
Comment (strager) [reply test/test-lex.cpp:1880]: I meant test `\ufeffStuff`, not `#!Stuff`.

Comment (singalhimanshu) [reply test/test-lex.cpp:1880]: oh okay, I misread that.

Comment (singalhimanshu) [reply test/test-lex.cpp:1880]: > I meant test `\ufeffStuff`, not `#!Stuff`.

Isn't this function testing that already - https://github.com/quick-lint/quick-lint-js/blob/6c54dc042d691bbbe6d0c83115b5cd6e907c92bd/test/test-lex.cpp#L1781

Let me know if I'm wrong.

File: docs/errors/E095.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# E095: unicode byte order mark (BOM) cannot appear before #! at beginning of script
+
+A script cannot have unicode byte order mark (BOM) before `#!`
+
+    #!/usr/bin/env node
+    let x = 3;
+    console.log(x);
+
+To fix this error, remove the unicode BOM before `#!`
```
Comment (strager) [inline docs/errors/E095.md:9]: Nit: "Unicode" is capitalized:
```suggestion
A script cannot have Unicode byte order mark (BOM) before `#!`

    #!/usr/bin/env node
    let x = 3;
    console.log(x);

To fix this error, remove the Unicode BOM before `#!`
```

**Review by strager — APPROVED** [review]

File: docs/errors/E095.md (hunk 5 above)
Comment (singalhimanshu) [reply docs/errors/E095.md]: Done!

**Comment by singalhimanshu** [comment]
I think it is ready to merge

**Pull request merged**